### PR TITLE
pagination passes headers

### LINKED ISSUE #706: Recursive octokit.getNextPage(response) omits headers.
Hello `octokit/rest.js` maintainers!

It seems like recursive pagination with the `octokit.getNextPage(response)` method results in noninitial requests omitting some headers (the one that causes the error I'm getting is the missing User-Agent, although it looks like Authorization is also ommitted).

---

Please provide a minimal code sample to reproduce the issue if possible.

Using the [Pagination sample](https://github.com/octokit/rest.js#pagination) from the README:

```js
async function paginate (method) {
  let response = await method({per_page: 100});
  let {data} = response;
  while (client.hasNextPage(response)) {
    response = await client.getNextPage(response);
    data = data.concat(response.data);
  }
  return data;
}

paginate(client.users.getAll).then(data => {
  console.log(data.length);
});
```

throws the error (Octokit debug logs included):

```
octokit:rest REQUEST: Url {
  protocol: 'https:',
  slashes: true,
  auth: null,
  host: 'api.github.com',
  port: null,
  hostname: 'api.github.com',
  hash: null,
  search: '?per_page=100',
  query: 'per_page=100',
  pathname: '/users',
  path: '/users?per_page=100',
  href: 'https://api.github.com/users?per_page=100',
  method: 'get',
  body: undefined,
  headers: 
   { accept: 'application/vnd.github.v3+json',
     'user-agent': 'octokit/rest.js v14.0.3',
     authorization: '[REDACTED]' },
  ca: undefined,
  family: undefined,
  proxy: undefined,
  rejectUnauthorized: true,
  timeout: 0 } +0ms

  octokit:rest STATUS: 200 +665ms

  octokit:rest HEADERS: {"date":"Fri, 19 Jan 2018 03:03:56 GMT","content-type":"application/json; charset=utf-8","content-length":"86247","connection":"close","server":"GitHub.com","status":"200 OK","x-ratelimit-limit":"5000","x-ratelimit-remaining":"4980","x-ratelimit-reset":"1516333134","cache-control":"private, max-age=60, s-maxage=60","vary":"Accept, Authorization, Cookie, X-GitHub-OTP, Accept-Encoding","etag":"\"85b1f3e97952bba8b1614e3c777a43b8\"","x-github-media-type":"github.v3; format=json","link":"<https://api.github.com/users?per_page=100&since=135>; rel=\"next\", <https://api.github.com/users{?since}>; rel=\"first\"","access-control-expose-headers":"ETag, Link, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval","access-control-allow-origin":"*","content-security-policy":"default-src 'none'","strict-transport-security":"max-age=31536000; includeSubdomains; preload","x-content-type-options":"nosniff","x-frame-options":"deny","x-xss-protection":"1; mode=block","x-runtime-rack":"0.329031","x-github-request-id":"E0B0:1EFFA:241E194:2A5FCF4:5A61601C"} +0ms

  octokit:rest REQUEST: Url {
  protocol: 'https:',
  slashes: true,
  auth: null,
  host: 'api.github.com',
  port: null,
  hostname: 'api.github.com',
  hash: null,
  search: '?per_page=100&since=135',
  query: 'per_page=100&since=135',
  pathname: '/users',
  path: '/users?per_page=100&since=135',
  href: 'https://api.github.com/users?per_page=100&since=135',
  headers: { accept: 'application/vnd.github.v3+json' } } +84ms

  octokit:rest STATUS: 403 +161ms

  octokit:rest HEADERS: {"cache-control":"no-cache","connection":"close","content-type":"text/html"} +0ms

UNHANDLED PROMISE REJECTION: Promise {
  <rejected> { [Error: Request forbidden by administrative rules. Please make sure your request has a User-Agent header (http://developer.github.com/v3/#user-agent-required). Check https://developer.github.com for other possible causes.
]
  message: 'Request forbidden by administrative rules. Please make sure your request has a User-Agent header (http://developer.github.com/v3/#user-agent-required). Check https://developer.github.com for other possible causes.\n',
  code: 403,
  status: undefined,
  headers: 
   { 'cache-control': 'no-cache',
     connection: 'close',
     'content-type': 'text/html' } } }
```

### PR DESCRIPTION
**Pull request by gr2m**
closes #706

### CODE REVIEW

**Comment by gr2m** [comment]
@synicalsyntax could you please test if that PR fixes the pagination problem for you?

**Comment by synicalsyntax** [comment]
Have you considered passing the `octokit` object to `getPage.js` as `this` instead of a parameter? I think that might be a possible code improvement, but otherwise, the functionality is great, thanks! :)

File: lib/plugins/pagination/index.js [hunk 1]
```diff
@@ -1,10 +1,10 @@
 module.exports = paginationPlugin
 
 function paginationPlugin (octokit) {
-  octokit.getFirstPage = require('./get-first-page')
-  octokit.getLastPage = require('./get-last-page')
-  octokit.getNextPage = require('./get-next-page')
-  octokit.getPreviousPage = require('./get-previous-page')
+  octokit.getFirstPage = require('./get-first-page').bind(null, octokit)
```
Comment (synicalsyntax) [inline lib/plugins/pagination/index.js:4]: `.bind(octokit)`

File: lib/plugins/pagination/get-first-page.js [hunk 2]
```diff
@@ -2,6 +2,6 @@ module.exports = getFirstPage
 
 const getPage = require('./get-page')
 
-function getFirstPage (link, headers, callback) {
-  return getPage(link, 'first', headers, callback)
+function getFirstPage (octokit, link, headers, callback) {
```
Comment (synicalsyntax) [inline lib/plugins/pagination/get-first-page.js:5]:
```js
function getFirstPage (link, headers, callback) {
 return getPage.apply(this, [link, 'prev', headers, callback])
}
```

File: lib/plugins/pagination/get-page.js [hunk 3]
```diff
@@ -2,9 +2,8 @@ module.exports = getPage
 
 const errors = require('../../request/errors')
 const getPageLinks = require('./get-page-links')
-const request = require('../../request/request')
 
-function getPage (link, which, headers, callback) {
+function getPage (octokit, link, which, headers, callback) {
```
Comment (synicalsyntax) [inline lib/plugins/pagination/get-page.js:6]: `function getPage (link, which, headers, callback) {`

File: lib/plugins/pagination/get-page.js [hunk 4]
```diff
@@ -25,7 +24,7 @@ function getPage (link, which, headers, callback) {
     headers: applyAcceptHeader(link, headers)
   }
 
-  const promise = request(requestOptions)
+  const promise = octokit.request(requestOptions)
```
Comment (synicalsyntax) [inline lib/plugins/pagination/get-page.js:27]: `const promise = this.request(requestOptions)`

**Review by synicalsyntax — COMMENTED** [review]
To clarify on my previous suggestion, these would be the changes necessary to pass the `octokit` object to pagination methods under the `this` keyword.

**Comment by gr2m** [comment]
I try to avoid `this` altogether, it remains a big cause of confusion and hard-to-find bugs, hence the pre-binding of the `octokit` arguments.

Thanks for checking it, the release will go out shortly via semantic-release

**Comment by gr2m** [comment]
released as [v14.0.4](https://github.com/octokit/rest.js/releases/tag/v14.0.4)

**Pull request merged**